### PR TITLE
Refactor processVoltDBStatement() in DDLCompiler

### PR DESCRIPTION
**Pull request by yzhang1991**
Basically, this refactoring commit just split the logics for processing VoltDB statements into separate classes and chained them up.

```java
m_statementProcessor.addNextProcessorFor(new CreateProcedureFromClass(this))
                    .addNextProcessorFor(new CreateProcedureAsScript(this))
                    .addNextProcessorFor(new CreateProcedureAsSQL(this))
                    .addNextProcessorFor(new CreateFunctionFromMethod(this))
                    .addNextProcessorFor(new DropFunction(this))
                    .addNextProcessorFor(new DropProcedure(this))
                    .addNextProcessorFor(new PartitionStatement(this))
                    .addNextProcessorFor(new ReplicateTable(this))
                    .addNextProcessorFor(new ImportClass(this))
                    .addNextProcessorFor(new CreateRole(this))
                    .addNextProcessorFor(new DropRole(this))
                    .addNextProcessorFor(new DropStream(this))
                    .addNextProcessorFor(new DRTable(this))
                    .addNextProcessorFor(new SetGlobalParam(this))
                    // CatchAllStatement need to be the last processor in the chain.
                    .addNextProcessorFor(new CatchAllStatement(m_statementProcessor));


private boolean processVoltDBStatement(DDLStatement ddlStatement, Database db,
                                       DdlProceduresToLoad whichProcs)
        throws VoltCompilerException
{
    return m_statementProcessor.process(ddlStatement, db, whichProcs);
}
```

### CODE REVIEW

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 1]
```diff
@@ -181,6 +183,107 @@ public DDLCompiler(VoltCompiler compiler,
...
+                            .addNextProcessorFor(new CreateRole(this))
+                            .addNextProcessorFor(new DropRole(this))
+                            .addNextProcessorFor(new DropStream(this))
+                            .addNextProcessorFor(new DRTable(this))
+                            .addNextProcessorFor(new SetGlobalParam(this))
+                            // CatchAllStatement need to be the last processor in the chain.
+                            .addNextProcessorFor(new CatchAllStatement(this, m_statementProcessor));
+    }
+
+    public static abstract class StatementProcessor {
```
Comment (wolffcm) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: So `StatementProcessor` is an abstract class that implements a linked list, where the elements are all instances of the classes in the `statements` sub-package?   That should be stated in a comment somewhere.
It seems strange to me that an abstact class would be a nested in another class.  Maybe it's not that strange.

File: src/frontend/org/voltdb/compiler/VoltCompiler.java [hunk 2]
```diff
@@ -221,7 +221,7 @@ public String getMessage() {
         }
     }
 
-    class VoltCompilerException extends Exception {
+    public class VoltCompilerException extends Exception {
```
Comment (wolffcm) [inline src/frontend/org/voltdb/compiler/VoltCompiler.java]: These all have to be public now?  I would think that they're being accessed from `org.voltdb.compiler.statements`, is package visibility not sufficient?  I'm not sure, that's why I'm asking.

File: src/frontend/org/voltdb/compiler/statements/CatchAllStatement.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import org.voltdb.catalog.Database;
+import org.voltdb.compiler.DDLCompiler;
+import org.voltdb.compiler.DDLCompiler.DDLStatement;
+import org.voltdb.compiler.DDLCompiler.StatementProcessor;
+import org.voltdb.compiler.VoltCompiler.DdlProceduresToLoad;
+import org.voltdb.compiler.VoltCompiler.VoltCompilerException;
+
+public class CatchAllStatement extends StatementProcessor {
+
+    private VoltDBStatementProcessor m_firstProcessor;
```
Comment (wolffcm) [inline src/frontend/org/voltdb/compiler/statements/CatchAllStatement.java]: What is the first processor?  The first processor that failed?

File: src/frontend/org/voltdb/compiler/statements/VoltDBStatementProcessor.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.regex.Matcher;
+
+import org.voltdb.catalog.Database;
+import org.voltdb.compiler.DDLCompiler;
+import org.voltdb.compiler.DDLCompiler.DDLStatement;
+import org.voltdb.compiler.DDLCompiler.StatementProcessor;
+import org.voltdb.compiler.VoltCompiler.DdlProceduresToLoad;
+import org.voltdb.compiler.VoltCompiler.VoltCompilerException;
+import org.voltdb.parser.SQLParser;
+
```
Comment (wolffcm) [inline src/frontend/org/voltdb/compiler/statements/VoltDBStatementProcessor.java]: Would be good to have a header comment here, and in all the new classes you've added.

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 5]
```diff
@@ -181,6 +183,107 @@ public DDLCompiler(VoltCompiler compiler,
...
+        private StatementProcessor m_next;
+
+        protected HSQLInterface m_hsql;
+        protected VoltCompiler m_compiler;
+        protected VoltDDLElementTracker m_tracker;
+        protected ClassLoader m_classLoader;
+        protected VoltXMLElement m_schema;
+        protected ClassMatcher m_classMatcher;
+        protected boolean m_isFinal = false;
```
Comment (wolffcm) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: What does `m_isFinal` do?  There should be a comment describing it.

File: src/frontend/org/voltdb/compiler/statements/CatchAllStatement.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    protected boolean processStatement(DDLStatement ddlStatement, Database db, DdlProceduresToLoad whichProcs)
+            throws VoltCompilerException {
+        /*
+         * if no correct syntax regex matched above then at this juncture
+         * the statement is syntax incorrect
+         */
+        if (PARTITION.equals(m_firstProcessor.getCommandPrefix())) {
+            throw m_compiler.new VoltCompilerException(String.format(
```
Comment (wolffcm) [inline src/frontend/org/voltdb/compiler/statements/CatchAllStatement.java]: Seems like these statement-specific error messages could go in the StatementProcessor classes?

File: src/frontend/org/voltdb/compiler/VoltCompiler.java (hunk 2 above)
Comment (yzhang1991) [inline src/frontend/org/voltdb/compiler/VoltCompiler.java]: Not sufficient.

**Comment by yzhang1991** [comment]
This will go into master.

**Pull request closed without merging**